File: project/includes/shapes.cpp
#include "shapes.hpp"

// draws a polygon over the image
void shapes::draw_polygon(
       cv::Mat img,
       std::vector<cv::Point> corners,
       cv::Scalar color,
       int thickness,
       int type,
       int shift
){

    unsigned int s = corners.size();
    if(s < 2) return;

    cv::line(img, corners[s-1], corners[0], color, thickness, type, shift);

    unsigned int i;
    for(i=1; i<s; i++){
        cv::line(img, corners[i-1], corners[i], color, thickness, type, shift);
    }
}

// create points for a trapazoid
std::vector<cv::Point> shapes::trapazoid(
    int a,
    int b,
    int center,
    int height,
    int y
){
    std::vector<cv::Point> corners;
    corners.push_back(cv::Point(center - a, y));          // point1
    corners.push_back(cv::Point(center + a, y));          // point2
    corners.push_back(cv::Point(center + b, y + height)); // point3
    corners.push_back(cv::Point(center - b, y + height)); // point4
    return corners;
}

// create points for a rectangle
std::vector<cv::Point> shapes::rectangle( int height, int width ){
    std::vector<cv::Point> corners;
    corners.push_back(cv::Point(     0 ,      0 )); // point1
    corners.push_back(cv::Point( width ,      0 )); // point2
    corners.push_back(cv::Point( width , height )); // point3
    corners.push_back(cv::Point(     0 , height )); // point4
    return corners;
}

// create points for a rectangle
void shapes::plot(
    cv::Mat canvas,
    cv::Mat vals_org,
    cv::Scalar color,
    int thickness,
    int y_max
){
    // clone vals
    cv::Mat vals(vals_org);

    if(y_max > -1){
        //printf("y_max: %d\n", y_max);

        vals.at<float>(0) = y_max;
        for( int i = 1; i < vals.rows; i++ ){
            if(vals.at<float>(i) > y_max){
                vals.at<float>(i) = y_max;
            }
        }


        //float sushi = 0;
        //for( int i = 0; i < vals.rows; i++ ){
        //    printf("%d -> %f\n", i, vals.at<float>(i));
        //    //sushi += vals.at<float>(i);
        //}
        //printf("sushi: %f\n", sushi);
    }

    // prepare unit width for x-axis streching
    int unit_w = cvRound( (double) canvas.cols/vals.rows );

    // normalize values for y-axis stretching
    cv::normalize(vals, vals, 0, canvas.rows, cv::NORM_MINMAX, -1, cv::Mat() );

    for( int i = 2; i < vals.rows; i++ ){
        cv::line( canvas, cv::Point( unit_w*(i-1), canvas.rows - cvRound(vals.at<float>(i-1)) ) ,
                          cv::Point( unit_w*(i)  , canvas.rows - cvRound(vals.at<float>(i))   ) ,
                          color, thickness, 8, 0  );
    }

}

File: project/python/shapes.py
import os
import numpy as np
import cv2

def trapazoid(a, b, center, height, y):
    return np.matrix([[int(center - a), int(         y)],  # point1
                      [int(center + a), int(         y)],  # point2
                      [int(center + b), int(y + height)],  # point3
                      [int(center - b), int(y + height)]]) # point4

def rectangle(height, width):
    return np.matrix([[ int(     0) , int(    0) ],  # point1
                      [ int(height) , int(    0) ],  # point2
                      [ int(height) , int(width) ],  # point3
                      [ int(     0) , int(width) ]]) # point4

def draw_polygon(img, fig, color=(255,255,255), thickness=1, t=8, shift=0, close=True):
    points = np.matrix(fig)

    if len(points) <= 1:
        return

    if close==True:
        points = np.vstack([points, points[0]])

    a = points[0]
    for b in points[1:]:
        cv2.line(img, (int(a[0,0]), int(a[0,1])), (int(b[0,0]), int(b[0,1])), color, thickness)
        a = b

def fill_polygon(img, fig, color=(255,255,255)):
    points = np.matrix(fig)
    cv2.fillPoly(img, pts=[points], color=color)

def plot(canvas, plot, y_max=None, color=(255,0,0), thickness=2):

    h = canvas.shape[0]
    w = canvas.shape[1]

    if y_max is None:
        y_max = np.max(plot)

    plot = plot/y_max * h

    l = len(plot)
    bw = w/l

    for x in range(l-1):
        v = plot[x]

        p1 = ( int(    x*bw) , int(  h - plot[x]) )
        p2 = ( int((x+1)*bw) , int(h - plot[x+1]) )

        cv2.line(canvas, p1, p2, color, thickness)
